In [1]:
# Grid search for findfing the best hyperparameters

In [2]:
from sklearn.datasets import load_digits
from sklearn.model_selection import GridSearchCV
from sklearn import svm

In [3]:
params = [
    {
        'kernel': ['linear','rbf','poly','sigmoid'],
        'C': [0.1,0.2,0.5,1.0,2.0,5.0]
    }
]

# Dataset Preparation

In [4]:
data = load_digits()
X = data.data
Y = data.target
print(X.shape, Y.shape)

(1797, 64) (1797,)


## Logistic Regression

In [6]:
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import cross_val_score

In [14]:
lr = LogisticRegression()
# lr.fit(X,Y)
cross_val_score(lr,X,Y,scoring="accuracy",cv=5).mean()

/home/ankan/Workstation/DataScience/jupyterenv/lib/python3.8/site-packages/sklearn/linear_model/_logistic.py:444: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
/home/ankan/Workstation/DataScience/jupyterenv/lib/python3.8/site-packages/sklearn/linear_model/_logistic.py:444: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.or

0.9137650882079852

In [15]:
## Logistic regression gives 92% accuracy

In [16]:
from sklearn import svm


In [22]:
svc = svm.SVC()
cross_val_score(svc,X,Y,scoring="accuracy",cv=5).mean()

0.9632838130609718

In [23]:
gs = GridSearchCV(estimator=svm.SVC(),param_grid=params, scoring="accuracy", cv=5, n_jobs=4)

In [24]:
gs.fit(X,Y)

GridSearchCV(cv=5, estimator=SVC(), n_jobs=4,
             param_grid=[{'C': [0.1, 0.2, 0.5, 1.0, 2.0, 5.0],
                          'kernel': ['linear', 'rbf', 'poly', 'sigmoid']}],
             scoring='accuracy')

In [25]:
gs.best_estimator_

SVC(C=5.0)

In [26]:
gs.best_params_

{'C': 5.0, 'kernel': 'rbf'}

In [27]:
gs.best_score_

0.9738502011761063

# SVM with rbf kernel and C=5 gives 97% accuracy on the dataset